### PR TITLE
Description and example of using the ESC command 'custom'

### PR DESCRIPTION
**Pull request by BHSDuncan**
…stom'.

### CODE REVIEW

File: advanced/faq.rst [hunk 1]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
 they will cause problems with mouse interactions. You will need to
 modify the properties of the ``TextureRect`` / ``ColorRect`` and set its
 "Mouse Filter" setting to "Ignore".
+
+I want to call my own GDScript function from ESC. How can I do this?
+--------------------------------------------------------------------
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
How can I call my own GDScript function from Escoria?
-----------------------------------------------------
```

File: advanced/faq.rst [hunk 2]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+ESC makes this possible by providing the `custom` command. The function
+must be associated with the child node of an `ESCItem` in your scene.
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
Escoria script makes this possible by providing the `custom` command. The function
must be associated with the child node of an `ESCItem` in your scene.
```

File: advanced/faq.rst [hunk 3]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+(See :doc:`https://docs.escoria-framework.org/en/devel/scripting/z_esc_reference.html#custom-object-node-func-name-params-api-doc`)
+
+To make use of this handy command, follow these steps:
+
+* Create a new node of the appropriate type as a child of an `ESCItem` in your
+  scene. For example, if you have a door in your scene and you want to make a
+  function that does something special when the door opens, you could create a
+  child node of type `Node2D` underneath the `ESCItem` that represents that
+  door.
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
* Create a new node of an appropriate type as a child of an `ESCItem` in your
  scene. For example, if you have a door in your scene and you want to make a
  function that does something special when the door opens, you could create a
  child node of type `Node2D` underneath the `ESCItem` that represents the
  door.
```

File: advanced/faq.rst [hunk 4]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+.. hint::
+
+   Remember the global ID of this `ESCItem`. You'll need it when you use the
+   `custom` command.
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
.. hint::

   Remember the global ID and name of this `ESCItem`. 
   You'll need them when you use the `custom` command.
```

File: advanced/faq.rst [hunk 5]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+* Create your GDScript function in a .gd file and attach it to the child node
+  as described above. This function can be called anything you want, but it
+  must take exactly one argument. Escoria will pass in to the function any and
+  all arguments pass in via the `custom` command as an array. If your function
+  doesn't require any arguments, Escoria will pass in an empty array.
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
* Create your GDScript function in a .gd file and attach it to the child node
  as described above. This function can be called anything you want, but it
  **must** take exactly one argument. Escoria will pass in to the function any and
  all arguments specified in the `custom` command as an array. If your function
  doesn't require any arguments, Escoria will pass in an empty array.
```

File: advanced/faq.rst [hunk 6]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+.. hint::
+
+   Remember the name of the node that makes use of your script. You'll need it
+   when you use the `custom` command.
+
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
```

File: advanced/faq.rst [hunk 7]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+.. hint::
+
+  It is up to you to unpack the arguments passed in via the `custom` command as
+  well as to perform any validation on these arguments you deem necessary.
```
Comment (balloonpopper) [inline advanced/faq.rst:476]:
```suggestion
  It is up to you to unpack the arguments passed in via the `custom` command as
  well as performing any validation on these arguments you deem necessary.
```

File: advanced/faq.rst [hunk 8]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+* In the appropriate event in an ESC script file, call the function you made
+  above by using the `custom` command, e.g.
+
+.. code-block::
+
+   :setup
+   # Call the function you defined with some arguments
+   custom esc_item_global_id child_node_name function_name arg1 arg2 arg3
+
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
* In the appropriate event in an Escoria script file, call the function you made
  above by using the `custom` command. For example :
     * Your game has a node with a global_id of "treasure_maps"
     * The "treasure_maps" node a child node called "france_treasure_map".
     * "france_treasure_map" has a Godot script (in GDscript format) attached. 
        In this script is a function "draw_treasure" to show a map on the screen.
       In your Escoria script, you would call the script using :

.. code-block::

   :setup
   # Call the function you defined with some arguments
   # Format : custom esc_item_global_id child_node_name function_name arg1 arg2 ... 
   # Note no parameters are passed to the "draw_treasure" function
   custom treasure_maps france_treasure_map draw_treasure 
   
     * The GDScript it calls might look like this :

.. code-block::

   extends Node2D

   # An empty array is passed as a parameter to the function so we ignore it
   func draw_treasure(_ignored_parameter):
	get_node("../france_map_sprite").visible = true

* If instead you wanted to pass this script some parameters:
   * If your script took an x and y coordinate to display an "X marks the spot"
      graphic on the treasure map, your Escoria script code might look like :

.. code-block::

   :setup
   custom treasure_maps france_treasure_map draw_treasure xcoordinate ycoordinate

   * The matching GDScript code now needs to extract those coordinates from the 
      parameter array that is passed to the function.

.. code-block::

   extends Node2D

   # Extract coordinates from the array of parameters
   func draw_treasure(list_of_parameters):
	var xcoord = list_of_parameters[0]
	var ycoord = list_of_parameters[1]
	get_node("../france_map_sprite").visible = true
	[...]

```

File: advanced/faq.rst [hunk 9]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+.. hint::
+
+   You can pass in as many arguments you want to your function. If no arguments
+   are required, you don't need to specify anything else after the function's
+   name.
+
+.. hint::
+
+   Remember that any arguments you pass to your function must be literals.
+   Variables like ESC flags, globals and the like cannot be used currently.
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
.. hint::

   Any arguments you pass to your function must be literals.
   Escoria variables (ESC flags, globals, etc) cannot be passed currently.
```

File: advanced/faq.rst [hunk 10]
```diff
@@ -441,3 +441,65 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+
+Now, when the above `:setup` event is run, it will call `function_name` with
+the specified arguments
+
+That's all there is to it!
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
```

**Review by balloonpopper — CHANGES_REQUESTED** [review]
I updated the language a little to make it consistent with the rest of the FAQ (the format of the question, and "ESC" to "Escoria script" for example). I also added a concrete example so users have something to copy.

File: advanced/faq.rst (hunk 3 above)
Comment (BHSDuncan) [reply advanced/faq.rst]: Frankly, in this context, they're interchangeable, but for the sake of consistency (given that you've done more work on the docs now than I have), I'll use your suggestion.

File: advanced/faq.rst (hunk 7 above)
Comment (BHSDuncan) [reply advanced/faq.rst:476]: Disagree here. While your suggestion does sound good, consistency between the two instructions and the preceding subject has to be maintained. 

I.e. if the first item wasn't there (about unpacking), we'd say, "It is up to you to perform any validation...", not "It is up to you performing validation...".

However, if you feel strongly enough... :)

File: advanced/faq.rst [hunk 11]
```diff
@@ -441,3 +441,96 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+* If instead you wanted to pass this script some parameters:
+   * If your script took an x and y coordinate to display an "X marks the spot"
+      graphic on the treasure map, your Escoria script code might look like :
+
+.. code-block::
+
+   :setup
+   custom treasure_maps france_treasure_map draw_treasure xcoordinate ycoordinate
+
+   * The matching GDScript code now needs to extract those coordinates from the
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion

.. hint::

   Escoria variables (ESC flags, globals, etc) can be passed as 
   parameters by wrapping them in braces.
   e.g. {xcoorinate}

   * The matching GDScript code now needs to extract those coordinates from the 
```

File: advanced/faq.rst [hunk 12]
```diff
@@ -441,3 +441,96 @@ If you use a control node like ``TextureRect`` or ``ColorRect``,
...
+	get_node("../france_map_sprite").visible = true
+
+* If instead you wanted to pass this script some parameters:
+   * If your script took an x and y coordinate to display an "X marks the spot"
+      graphic on the treasure map, your Escoria script code might look like :
+
+.. code-block::
+
+   :setup
+   custom treasure_maps france_treasure_map draw_treasure xcoordinate ycoordinate
```
Comment (balloonpopper) [inline advanced/faq.rst]:
```suggestion
   custom treasure_maps france_treasure_map draw_treasure {xcoordinate} {ycoordinate}
```

**Comment by balloonpopper** [comment]
Added 2 changes based on https://github.com/godot-escoria/escoria-demo-game/pull/593

**Review by balloonpopper — APPROVED** [review]

**Review by StraToN — APPROVED** [review]
No problem with the text. Just the doc8 fixes to be done. Thanks!

**Comment by StraToN** [comment]
Thanks a lot!

**Pull request merged**